plugins/inputs/t128_tank/t128_tank.go

```diff
@@ -0,0 +1,114 @@
+package t128_tank
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+const (
+	T128TankAddr = "127.0.0.1"
+	T128TankPort = 11011
+)
+
+var sampleConfig = `
+[[inputs.t128_tank]]
+## A (unique) file to use for index tracking. 
+## This tracking allows each event to be produced once.
+# index_file = ""
+
+## Required. The TANK topic to consume.
+# topic = "events"
+
+## Port Number to get tank data from.
+# port_number = 11011
+
+## Server Address to get tank data from.
+# server_address = "127.0.0.1"
+`
+
+type T128Tank struct {
+	IndexFile     string `toml:"index_file"`
+	Topic         string `toml:"topic"`
+	PortNumber    int    `toml:"port_number"`
+	ServerAddress string `toml:"server_address"`
+	Log           telegraf.Logger
+	ctx           context.Context
+	cancel        context.CancelFunc
+}
+
+func (*T128Tank) SampleConfig() string {
+	return sampleConfig
+}
+
+func (*T128Tank) Description() string {
+	return "Run TANK as a long-running input plugin"
+}
+
+func (plugin *T128Tank) Init() error {
+	err := plugin.checkConfig()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (plugin *T128Tank) Gather(_ telegraf.Accumulator) error {
+	return nil
+}
+
+func (plugin *T128Tank) Start(acc telegraf.Accumulator) error {
+	plugin.ctx, plugin.cancel = context.WithCancel(context.Background())
+
+	reader := NewReader(plugin.ServerAddress, plugin.PortNumber, plugin.Topic, plugin.IndexFile, StartIndex, plugin.Log, plugin.ctx, acc)
+	go func() {
+		for {
+			select {
+			case <-plugin.ctx.Done():
+				return
+			case messages := <-reader.sendChan:
+				for _, message := range messages {
+					acc.AddFields("t128_tank", map[string]interface{}{
+						"message": string(message.Message),
+					}, map[string]string{
+						"index": strconv.FormatUint(message.Index.value, 10),
+					}, time.Now())
+				}
+			}
+		}
+	}()
+
+	go func() {
+		reader.Run()
```

Comment:
not too important, but typically the context would be passed to `Run(ctx)` instead of to the reader constructor. In theory the reader could be `Run` again, but as written that's not possible.

Reply:
gotcha! changing them